test/kubetest/oidc.go

```diff
@@ -0,0 +1,244 @@
+limitations under the License.
+*/
+package kubetest
+
+import (
+	"context"
+	"crypto/sha1"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/json"
+	"fmt"
+	mockserver "github.com/BraspagDevelopers/mock-server-client"
+	"gopkg.in/go-jose/go-jose.v2"
+	"gopkg.in/square/go-jose.v2/jwt"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"net/http"
+	"time"
+)
+
+const (
+	DefaultCertsURL  = "/certs"
+	DefaultConfigURL = "/.well-known/openid-configuration"
+
+	defaultAlgorithm = "RS256"
+	defaultAudience  = "test-client-id"
+	defaultClientID  = "test-client-id"
+	defaultIssuer    = "https://mock-issuer.default.svc.cluster.local:8443"
+	defaultKeyID     = "01234567890"
+	defaultRole      = "metrics"
+	defaultUse       = "sign"
+	defaultUsername  = "test-client"
+)
+
+// CustomClaims defines additional claims that need to be added to the JWT to test the functionality supported by the
+// proxy.
+type CustomClaims struct {
+	*jwt.Claims
+	Username string   `json:"preferred_username,omitempty"`
+	Roles    []string `json:"roles,omitempty"`
+}
+
+// OIDCTokenConfig contains the data required to create a new token and the associated materials.
+type OIDCTokenConfig struct {
+	Algorithm      string
+	Audience       []string
+	ExpiresAt      time.Time
+	IssuedAt       time.Time
+	Issuer         string
+	KeyID          string
+	Name           string
+	PublishedKeyID *string
+	Roles          []string
+	Subject        string
+	Use            string
+	Username       string
+}
+
+// NewTokenConfig creates a token
+func NewTokenConfig(name string) *OIDCTokenConfig {
+	return &OIDCTokenConfig{
+		Algorithm: defaultAlgorithm,
+		Audience:  []string{defaultAudience},
+		ExpiresAt: time.Now().Add(30 * time.Minute),
+		IssuedAt:  time.Now(),
+		Issuer:    defaultIssuer,
+		KeyID:     defaultKeyID,
+		Name:      name,
+		Roles:     []string{defaultRole},
+		Subject:   defaultClientID,
+		Use:       defaultUse,
+		Username:  defaultUsername,
+	}
+}
+
+// CreateOIDCIssuer inserts expectations into a mock-server to act as a basic OIDC issuer
+func CreateOIDCIssuer(issuerURL, mockServerURL string) Action {
+	return func(ctx *ScenarioContext) error {
+		mockClient := mockserver.NewClientURL(mockServerURL)
+		err := mockClient.RegisterExpectation(mockserver.Expectation{
+			Request: mockserver.RequestMatcher{
+				Method: http.MethodGet,
+				Path:   DefaultConfigURL,
+				Body: mockserver.BodyMatcher{
+					Type:      mockserver.MatchBodyJSON,
+					JSON:      nil,
+					MatchType: mockserver.TolerantMatch,
+				},
+			},
+			Response: mockserver.Response{
+				Body: map[string]interface{}{
+					"issuer":                                issuerURL,
+					"authorization_endpoint":                fmt.Sprintf("%s/authorize", issuerURL),
+					"token_endpoint":                        fmt.Sprintf("%s/token", issuerURL),
+					"jwks_uri":                              fmt.Sprintf("%s%s", issuerURL, DefaultCertsURL),
```

Comment:
Ya, that's an unfortunate choice.  ... Keycloak maps it to `/certs` so I must have focused on that part instead of the `jwks_uri` attribute name.  I'll update.